src/textual/_on.py

```diff
@@ -0,0 +1,56 @@
+from __future__ import annotations
+
+from typing import Callable, TypeVar
+
+from .css.parse import parse_selectors
+from .css.tokenizer import TokenError
+from .message import Message
+
+DecoratedType = TypeVar("DecoratedType")
+
+
+class OnDecoratorError(Exception):
+    """The selector in the 'on' decorator failed to parse."""
+
+
+def on(
+    message_type: type[Message], selector: str | None = None
+) -> Callable[[DecoratedType], DecoratedType]:
+    """Decorator to declare method is a message handler.
+
+    Example:
+        ```python
+        @on(Button.Pressed, "#quit")
+        def quit_button(self) -> None:
+            self.app.quit()
+        ```
+
+    Args:
+        message_type: The message type (i.e. the class).
+        selector: An optional [selector](/guide/CSS#selectors). If supplied, the handler will only be called if `selector`
+            matches the sender of the message.
```

Comment:
Might be a good idea to note that this sender is checked in the attribute `control`.